### PR TITLE
MT-hardening builder.ClasspathJar.knownPackageNames #3250

### PR DESCRIPTION
**Pull request by jukzi**
* use Set<String>
* cache a unmodifiable copy
* simplify PackageCacheEntry to record
* add missing final/volatile modifiers
* remove ClasspathJrt.PackageCache (only module names had been used)
* don't calculate unused location of module.info
* remove closeZipFileAtEnd (was always true when zipfile!=null)

https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3250

### CODE REVIEW

File: org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/ClasspathJrtWithReleaseOption.java [hunk 1]
```diff
@@ -71,13 +67,6 @@ public ClasspathJrtWithReleaseOption(String zipFilename, AccessRuleSet accessRul
 			this.externalAnnotationPath = externalAnnotationPath.toString();
 		}
 		this.release = getReleaseOptionFromCompliance(release);
-		JrtFileSystem systemForRelease = null;
```
[the remark this answers is not in the document]
Comment (jukzi) [reply org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/ClasspathJrtWithReleaseOption.java:74]: Compiler complained the code is unused after refactoring (to compute only modules not packages)

**Comment by jukzi** [comment]
> should be green again

tests failed because i used `modulesCache.keySet()`, instead of `modulesCache.get(jrt.getKey()).keySet()` in getModuleNames(). Unfortunatly both are `Set<String>` now, while one contains the jrt filenames and ther later the modulenames

**Comment by jukzi** [comment]
If anybody wants to review, please do it by Monday or leave a message.

File: org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/ClasspathJar.java [hunk 2]
```diff
@@ -49,80 +51,60 @@
...
-		readJarContent(packageSet);
-		return new PackageCacheEntry(this.zipFile, timestamp, this.fileSize, packageSet);
+		return new PackageCacheEntry(new WeakReference<>(this.zipFile), timestamp, this.fileSize, Set.copyOf(readPackageNames()));
 	});
 
 	return entry.packageSet;
 }
-protected String readJarContent(final SimpleSet packageSet) {
-	String modInfo = null;
+/** overloaded */
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/ClasspathJar.java:97]: @jukzi  this comment looks plain wrong.

Additionally a reference to the issue / PR is missing from the commit comment.

**Pull request merged**